File: k8-quote-api/src/main/java/com/quote/k8/service/QuoteManagementService.java
package com.quote.k8.service;

import com.quote.k8.dto.QuotePageResponse;
import com.quote.k8.model.Quote;
import com.quote.k8.repository.QuoteRepository;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.jboss.logging.Logger;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@ApplicationScoped
public class QuoteManagementService {

    private static final Logger LOG = Logger.getLogger(QuoteManagementService.class);

    @Inject
    QuoteRepository quoteRepository;

    public QuotePageResponse getQuotes(int page, int pageSize, String quoteText, String author, String sortBy, String sortOrder) {
        LOG.info("Getting quotes with pagination - page: " + page + ", pageSize: " + pageSize + 
                 ", sortBy: " + sortBy + ", sortOrder: " + sortOrder);

        List<Quote> allQuotes = quoteRepository.findAllQuotes();

        // Apply filters
        if (quoteText != null && !quoteText.isBlank()) {
            allQuotes = allQuotes.stream()
                    .filter(q -> q.quoteText != null && q.quoteText.toLowerCase().contains(quoteText.toLowerCase()))
                    .collect(Collectors.toList());
        }

        if (author != null && !author.isBlank()) {
            allQuotes = allQuotes.stream()
                    .filter(q -> q.author != null && q.author.toLowerCase().contains(author.toLowerCase()))
                    .collect(Collectors.toList());
        }

        // Apply sorting
        if (sortBy != null && !sortBy.isBlank()) {
            Comparator<Quote> comparator;
            boolean descending = sortOrder != null && sortOrder.equalsIgnoreCase("desc");

            switch (sortBy.toLowerCase()) {
                case "author":
                    comparator = Comparator.comparing(q -> q.author != null ? q.author.toLowerCase() : "");
                    break;
                case "created":
                    comparator = Comparator.comparing(q -> q.createdAt != null ? q.createdAt : java.time.LocalDateTime.MIN);
                    break;
                case "likes":
                    comparator = Comparator.comparing(q -> q.likeCount != null ? q.likeCount : 0);
                    break;
                case "id":
                default:
                    comparator = Comparator.comparing(q -> q.quoteId != null ? q.quoteId : 0);
                    break;
            }

            if (descending) {
                comparator = comparator.reversed();
            }

            allQuotes = allQuotes.stream()
                    .sorted(comparator)
                    .collect(Collectors.toList());
        } else {
            // Default sort by quoteId ascending
            allQuotes = allQuotes.stream()
                    .sorted(Comparator.comparing(q -> q.quoteId != null ? q.quoteId : 0))
                    .collect(Collectors.toList());
        }

        int totalCount = allQuotes.size();
        int totalPages = (int) Math.ceil((double) totalCount / pageSize);

        // Apply pagination
        int fromIndex = (page - 1) * pageSize;
        int toIndex = Math.min(fromIndex + pageSize, totalCount);
        List<Quote> paginatedQuotes = allQuotes.subList(fromIndex, toIndex);

        LOG.info("Returning " + paginatedQuotes.size() + " quotes (total: " + totalCount + ", pages: " + totalPages + ")");

        return new QuotePageResponse(paginatedQuotes, totalCount, page, pageSize, totalPages);
    }
}

File: doc/migration/quote-azure-k8-backend/Services/QuoteManagementService.cs
using Microsoft.Extensions.Logging;
using quote_azure_k8_backend.Models;
using quote_azure_k8_backend.Data;

namespace quote_azure_k8_backend.Services
{
    public class QuoteManagementService : IQuoteManagementService
    {
        private readonly IQuoteRepository _quoteRepository;
        private readonly IUserActivityRepository _userActivityRepository;
        private readonly IZenQuotesService _zenQuotesService;
        private readonly ILogger<QuoteManagementService> _logger;

        public QuoteManagementService(
            IQuoteRepository quoteRepository,
            IUserActivityRepository userActivityRepository,
            IZenQuotesService zenQuotesService,
            ILogger<QuoteManagementService> logger)
        {
            _quoteRepository = quoteRepository;
            _userActivityRepository = userActivityRepository;
            _zenQuotesService = zenQuotesService;
            _logger = logger;
        }

        public async Task<QuotePageResponse> GetQuotesAsync(int page, int pageSize, string? quoteText, string? author, string? sortBy, string? sortOrder)
        {
            try
            {
                var allQuotes = await _quoteRepository.GetAllQuotesAsync();
                
                // Apply filters
                if (!string.IsNullOrEmpty(quoteText))
                {
                    allQuotes = allQuotes.Where(q => q.QuoteText.Contains(quoteText, StringComparison.OrdinalIgnoreCase)).ToList();
                }
                
                if (!string.IsNullOrEmpty(author))
                {
                    allQuotes = allQuotes.Where(q => q.Author.Contains(author, StringComparison.OrdinalIgnoreCase)).ToList();
                }

                // Apply sorting
                if (!string.IsNullOrEmpty(sortBy))
                {
                    switch (sortBy.ToLower())
                    {
                        case "author":
                            allQuotes = sortOrder?.ToLower() == "desc" 
                                ? allQuotes.OrderByDescending(q => q.Author).ToList()
                                : allQuotes.OrderBy(q => q.Author).ToList();
                            break;
                        case "created":
                            allQuotes = sortOrder?.ToLower() == "desc"
                                ? allQuotes.OrderByDescending(q => q.CreatedAt).ToList()
                                : allQuotes.OrderBy(q => q.CreatedAt).ToList();
                            break;
                        case "likes":
                            allQuotes = sortOrder?.ToLower() == "desc"
                                ? allQuotes.OrderByDescending(q => q.LikeCount).ToList()
                                : allQuotes.OrderBy(q => q.LikeCount).ToList();
                            break;
                        default:
                            allQuotes = sortOrder?.ToLower() == "desc"
                                ? allQuotes.OrderByDescending(q => q.Id).ToList()
                                : allQuotes.OrderBy(q => q.Id).ToList();
                            break;
                    }
                }
                else
                {
                    allQuotes = allQuotes.OrderBy(q => q.Id).ToList();
                }

                var totalCount = allQuotes.Count;
                var totalPages = (int)Math.Ceiling((double)totalCount / pageSize);
                var quotes = allQuotes.Skip((page - 1) * pageSize).Take(pageSize).ToList();

                return new QuotePageResponse
                {
                    Quotes = quotes,
                    TotalCount = totalCount,
                    Page = page,
                    PageSize = pageSize,
                    TotalPages = totalPages
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting quotes with pagination");
                throw;
            }
        }

        public async Task<QuoteAddResponse> FetchAndAddNewQuotesAsync(string requestingUsername)
        {
            try
            {
                var newQuotes = await _zenQuotesService.GetMultipleQuotesAsync();
                var addedCount = 0;

                foreach (var quote in newQuotes)
                {
                    // Check if quote already exists (by text and author)
                    var existingQuotes = await _quoteRepository.GetAllQuotesAsync();
                    var exists = existingQuotes.Any(q => 
                        q.QuoteText.Equals(quote.QuoteText, StringComparison.OrdinalIgnoreCase) && 
                        q.Author.Equals(quote.Author, StringComparison.OrdinalIgnoreCase));

                    if (!exists)
                    {
                        // Assign new ID
                        var maxId = existingQuotes.Any() ? existingQuotes.Max(q => q.Id) : 0;
                        quote.Id = maxId + 1;
                        
                        await _quoteRepository.AddQuoteAsync(quote);
                        addedCount++;
                    }
                }

                var totalQuotes = (await _quoteRepository.GetAllQuotesAsync()).Count;

                _logger.LogInformation("Added {AddedCount} new quotes by {Username}", addedCount, requestingUsername);

                return new QuoteAddResponse
                {
                    QuotesAdded = addedCount,
                    TotalQuotes = totalQuotes,
                    Message = $"Successfully added {addedCount} new quotes"
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error fetching and adding new quotes");
                throw;
            }
        }

        public async Task<int> GetTotalQuotesCountAsync(string? quoteText = null, string? author = null)
        {
            try
            {
                var allQuotes = await _quoteRepository.GetAllQuotesAsync();
                
                if (!string.IsNullOrEmpty(quoteText))
                {
                    allQuotes = allQuotes.Where(q => q.QuoteText.Contains(quoteText, StringComparison.OrdinalIgnoreCase)).ToList();
                }
                
                if (!string.IsNullOrEmpty(author))
                {
                    allQuotes = allQuotes.Where(q => q.Author.Contains(author, StringComparison.OrdinalIgnoreCase)).ToList();
                }

                return allQuotes.Count;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting total quotes count");
                throw;
            }
        }

        public async Task<int> GetTotalLikesAsync()
        {
            try
            {
                return await _userActivityRepository.GetTotalLikesCountAsync();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting total likes count");
                throw;
            }
        }

        public async Task<Quote?> GetQuoteByIdAsync(int id)
        {
            try
            {
                return await _quoteRepository.GetQuoteByIdAsync(id);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting quote by ID: {QuoteId}", id);
                throw;
            }
        }

        public async Task<bool> DeleteQuoteAsync(int id, string requestingUsername)
        {
            try
            {
                var result = await _quoteRepository.DeleteQuoteAsync(id);
                
                if (result)
                {
                    _logger.LogInformation("Quote {QuoteId} deleted by {Username}", id, requestingUsername);
                }
                
                return result;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error deleting quote: {QuoteId}", id);
                throw;
            }
        }

        public async Task<Quote?> UpdateQuoteAsync(int id, Quote quote, string requestingUsername)
        {
            try
            {
                var existingQuote = await _quoteRepository.GetQuoteByIdAsync(id);
                if (existingQuote == null)
                    return null;

                existingQuote.QuoteText = quote.QuoteText;
                existingQuote.Author = quote.Author;
                existingQuote.Source = quote.Source;

                var updatedQuote = await _quoteRepository.UpdateQuoteAsync(existingQuote) ? existingQuote : null;
                
                if (updatedQuote != null)
                {
                    _logger.LogInformation("Quote {QuoteId} updated by {Username}", id, requestingUsername);
                }

                return updatedQuote;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating quote: {QuoteId}", id);
                throw;
            }
        }
    }
}
